Create rule S6522: No assigning values to variables used for imports (#1635)
SonarSource/rspec: rules/S6522/javascript/rule.adoc

Disallow assigning values to variables used for imports.

== Why is this an issue?

> Assigning a new value to an import variable is generally seen as a poor practice for readability.

Reviewer: ```suggestion
Assigning a new value to an import variable is a poor practice that affects code readability.
```

=== What is the potential impact?

Bad practices can cause bugs. When using the reassigned variable, if you are expecting to access the imported value, but are in fact using something else can lead to unexpected behaviour that is hard to debug.

> Additionnally, it will cause a runtime error when using ES modules. For typescript, it will be a comple error.

Reviewer: ```suggestion
More importantly, it will cause a runtime error when using ES modules. With TypeScript, it will raise a compilation error.
```

Reviewer: I am surprised that TypeScript raises a compilation error. Did you try that?

Reviewer 2: Yup:

```
error TS2632: Cannot assign to 'errorMiddleware' because it is an import.

33 errorMiddleware = 12;
```

== How to fix it

> You should use a separate variable to store your other values.

Reviewer: ```suggestion
You should use a separate variable to store your other values and make constant the import variable.
```

Reviewer 2: ...  when using CommonJS. ?

=== Code examples

==== Noncompliant code example

[source,javascript]
----

> const { readFile } = require('fs');

Reviewer: ```suggestion
let { readFile } = require('fs');
```

readFile = 'hello world!';
----

==== Compliant solution

[source,javascript]
----
const { readFile } = require('fs');
const text = 'hello world!';
----

//=== How does this work?

//=== Pitfalls

//=== Going the extra mile

//== Resources
//=== Documentation
//=== Articles & blog posts
//=== Conference presentations

> //=== Standards

Reviewer: Is it a new practice to comment those, or should we remove them altogether?
